### PR TITLE
[MODUSERBL-115, MODUSERBL-116]: Open transactions

### PR DESCRIPTION
**Pull request by rchr**
## What?

Added support to check if user has open transactions, and to delete a user (after checking for open transactions). Deletion will be executed iff user has no open transaction.

An open transaction is an open loan, open request, unpaid fee/fine, unexpired proxy, or manual block associated with the given user.

## Why?

As a service staff member with the proper permissions I want to be able to delete a user through the UI.

When the user has no open transactions attached to it, including open loans, open requests, open fees/fines, unexpired proxies or blocks, the user can be deleted. The list of relevant open transactions is defined in MODUSERBL-115.

Otherwise an information should be given with information about the open transactions  (UI details in UIU-1971). The user can be active or inactive.

In addition, it is needed to check if a user has open transaction (without deletion) via the API.

## How?

This PR adds three new API endpoints:

* GET: /bl-users/by-id/{id}/open-transactions

* GET: /bl-users/by-username/{name}/open-transactions

* DELETE: /bl-users/by-id/{id}

The two GET endpoints return a summary if there are any open transactions associated to the given user. I added the 'by-username' endpoint to be consistent with the GET endpoints `/bl-users/by-id/{id}` and `/bl-users/by-username/{name}`.

The DELETE endpoint checks if there are any open transactions associated to the user. If there are any it returns HTTP 409 Conflict status code with the list of transactions. If there aren't any transactions it calls the DELETE endpoint of mod-users.

## Tickets

https://issues.folio.org/browse/MODUSERBL-115
https://issues.folio.org/browse/MODUSERBL-116

### CODE REVIEW

File: README.md
Comment (marcjohnson-kint) [inline README.md:61]: Why are these endpoints needed?

Is the intention that they are used by the client prior to attempting deletion or after a failed deletion to present more information (rather than including it in the response of the failed deletion)?

File: descriptors/ModuleDescriptor-template.json
Comment (marcjohnson-kint) [inline descriptors/ModuleDescriptor-template.json:108]: The use of `manualblocks.collection.get` suggests that this endpoint depends upon a manual blocks API, and yet there seems to be no dependency for this in the `requires` section, is that intentional?

File: ramls/mod-users-bl.raml
Comment (marcjohnson-kint) [inline ramls/mod-users-bl.raml:188]: What responses are expected from a delete operation (there don't appear to be any documented here)?

**Comment by marcjohnson-kint** [comment]
@rchr Thank you for inviting me to review this pull request. I'm happy to provide some feedback.

Before I get to involved in doing so, please can you help me understand what kind of feedback you want from a review from 
me. For example, do you want feedback on the code / technical aspects of the change or on the design (which likely means I would need to understand the intent of the changes)?

I'm asking because no review is required in FOLIO's development process (which expects intra-team reviews only) which leads me to think you might want something specific.

File: README.md
Comment (marcjohnson-kint) [inline README.md:55]: Why is this line included for this endpoint and not the very similar one below?

**Comment by rchr** [comment]
@marcjohnson-kint Thank you for your comments so far!

I requested your review initially as I was unsure about the review process. However, as you said that only an intra-team review is required, you do not need to provide more feedback. I will wait for @alb3rtino's review.
Thanks again for your feedback! I will work on your comments.

File: README.md
Comment (rchr) [reply README.md:61]: The intention is that they are used prior to attempting deletion, e.g. via a campus identity management system when a user is leaving the university. The detailed use case is explained in [UXPROD-2904](https://issues.folio.org/browse/UXPROD-2904).

For a failed deletion additional information (an overview about open transactions) is already presented.

File: descriptors/ModuleDescriptor-template.json
Comment (rchr) [reply descriptors/ModuleDescriptor-template.json:108]: `manualblocks.collection.get` is defined by mod-feesfines. It is required for the `/manualblocks` endpoint that is provided by the interface `feesfines` `v16.3`. `feesfines` `v16.3` is already listed in the `requires` section.

File: README.md
Comment (rchr) [reply README.md:55]: Thanks for catching this! There was no intention behind this. I added the missing line to the similar ones below.

File: ramls/mod-users-bl.raml
Comment (rchr) [reply ramls/mod-users-bl.raml:165]: I remodelled the response to be more explicit now.

Before it was modelled as an optional HTTP method of a compositeUserResource, which was rather implicit.

**Review by alb3rtino — COMMENTED** [review]
I dont have anything else to add.

**Review by alb3rtino — APPROVED** [review]

File: README.md
Comment (marcjohnson-kint) [reply README.md]: Thanks for explaining that.

Personally, the name `transactions` makes me a little nervous, it seems a very general term to have defined. I'm aware that came from the PO asking for this change.

**Comment by marcjohnson-kint** [comment]
> However, as you said that only an intra-team review is required, you do not need to provide more feedback.

I'm happy to provide more feedback if you would like it. I have tried to get into the habit of asking folks why / what feedback they want because sometimes I give feedback that folks don't really want.

**Comment by marcjohnson-kint** [comment]
One thing I will add is that an impact of this change is that a minimal system (just enough to log in) will now require mod-circulation-storage and mod-feesfines.

The testing hosted reference environments enable modules in a specific order ([for example](https://github.com/folio-org/folio-ansible/blob/master/group_vars/testing-core)) and so this change will need coordinating with the DevOps folks in order to avoid outages to those environments.

**Comment by marcjohnson-kint** [comment]
Does this pull request cover the work for both MODUSERBL-115 and MODUSERBL-116?

File: descriptors/ModuleDescriptor-template.json
Comment (marcjohnson-kint) [reply descriptors/ModuleDescriptor-template.json:108]: Damn, thank you. I find it confusing that patron blocks are not implemented by the patron blocks module or in an interface that refers to patron blocks :-/

File: ramls/mod-users-bl.raml
Comment (marcjohnson-kint) [reply ramls/mod-users-bl.raml:165]: Thanks

**Comment by wafschneider** [comment]
I'm concerned about entangling interface dependencies in this manner. I understand that you need to have cross-module checks to ensure data integrity, but the cost is a continued reduction in the modularity of the system. Is there no way to implement this functionality using [optional interface dependencies](https://github.com/folio-org/okapi/blob/master/doc/guide.md#optional-interfaces) instead?

**Comment by marcjohnson-kint** [comment]
For context, it might be worth noting that the App Interaction SIG have been working on patterns in this area and the Technical Leads have an [ongoing conversation](https://wiki.folio.org/display/DD/Deletion+of+core-module+records+may+leave+dangling+references+from+non-core+modules) about how to general approach this kind of work

**Comment by rchr** [comment]
@marcjohnson-kint Yes, this PR covers both MODUSERBL-115 and MODUSERBL-116.

@wafschneider Thanks the hint to use optional interfaces. From a deployment perspective that makes sense to me. The implementation is also working with optional interfaces. Therefore, I switched the interfaces introduced in this PR (loan-storage and feesfines) to be optional interfaces.

**Comment by marcjohnson-kint** [comment]
> Yes, this PR covers both MODUSERBL-115 and MODUSERBL-116.

Thanks for confirming this. If you submit a single pull request for two issues, please can both be referenced in order to help with change and release tracking.

**Comment by rchr** [comment]
> How does the implementation behave when requests to those endpoints fail due to no provider?

Those requests behave as there are no requested entities for the current user, e.g. manualblocks.

The [okapi documentation](https://github.com/folio-org/okapi/blob/master/doc/guide.md#request-processing) says: "If no handlers are found, Okapi will return a 404 NOTFOUND.".

This PR handles 404 responses as empty optionals:
```java
...
case HttpStatus.SC_NOT_FOUND: // this is 404
            return Optional.empty();
...
```
This means, if we are requesting the _manualblocks_ of a user but the _feesfines_ interface is not existent, no manual blocks are reported for this user.

**Comment by marcjohnson-kint** [comment]
> > How does the implementation behave when requests to those endpoints fail due to no provider?
> 
> Those requests behave as there are no requested entities for the current user, e.g. manualblocks.
> 
> The [okapi documentation](https://github.com/folio-org/okapi/blob/master/doc/guide.md#request-processing) says: "If no handlers are found, Okapi will return a 404 NOTFOUND.".
> 
> This PR handles 404 responses as empty optionals (e.g. see [FeesFinesModuleClientImpl](https://github.com/folio-org/mod-users-bl/blob/8feed3e2e281b6fc16ea0fc3297375384fb9b910/src/main/java/org/folio/rest/client/impl/FeesFinesModuleClientImpl.java#L63)):
> 
> ```java
> ...
> case HttpStatus.SC_NOT_FOUND: // this is 404
>             return Optional.empty();
> ...
> ```
> 
> This means, if we are requesting the _manualblocks_ of a user but the _feesfines_ interface is not existent, no manual blocks are reported for this user.

Ok, we assume that that the only part of the system to respond with a 404 to those kinds of requests is Okapi when the endpoint isn't implemented?

**Comment by rchr** [comment]
> Ok, we assume that that the only part of the system to respond with a 404 to those kinds of requests is Okapi when the endpoint isn't implemented?

Yes. The endpoints to those kind of requests are modelled as resource types of type [collection](https://github.com/folio-org/raml/blob/raml1.0/rtypes/collection.raml) of FOLIO's [shared RAML and schema files](https://github.com/folio-org/raml).
In collection.raml a 404 response is not modelled.

Therefore I expect that the only part of the system to respond with a 404 to those kinds of requests is Okapi when the endpoint isn't implemented.

File: src/test/java/org/folio/rest/OpenTransactionsTest.java
Comment (julianladisch) [inline src/test/java/org/folio/rest/OpenTransactionsTest.java:68]: Race condition, insertData may start before verticle deployment has finished.
For details see
https://issues.folio.org/browse/MODUSERBL-118
https://github.com/folio-org/mod-users-bl/pull/117

File: ramls/mod-users-bl.raml
Comment (julianladisch) [inline ramls/mod-users-bl.raml:112]:
```suggestion
      description: "Check if user has any open transactions, and if so, how many. Identify user by <<typeName>>"
```

File: src/main/java/org/folio/service/transactions/OpenTransactionsServiceImpl.java
Comment (julianladisch) [inline src/main/java/org/folio/service/transactions/OpenTransactionsServiceImpl.java:40]: Why does it fetch loan records if all we need is the count? The loan records are fetched and then thrown away.
Consider using the `limit=0` feature of RMB that returns the exact (!) count without fetching any record:
https://github.com/folio-org/raml-module-builder#estimated-totalrecords

**Review by julianladisch — CHANGES_REQUESTED** [review]
See in-line comments.

File: ramls/mod-users-bl.raml
Comment (rchr) [reply ramls/mod-users-bl.raml]: Makes sense. I renamed both to openTransaction resp. openTransactions.json.

File: src/main/java/org/folio/service/transactions/OpenTransactionsServiceImpl.java
Comment (rchr) [reply src/main/java/org/folio/service/transactions/OpenTransactionsServiceImpl.java]: I switched to use RMB's `limit=0` feature for fetching the exact counts of open transactions.
I wasn't aware of that. Thanks for pointing me to that!

File: src/main/java/org/folio/rest/client/impl/CirculationStorageModuleClientImpl.java
Comment (rchr) [reply src/main/java/org/folio/rest/client/impl/CirculationStorageModuleClientImpl.java]: I switched to use RMB's `limit=0` feature (see comment below).

File: ramls/mod-users-bl.raml
Comment (rchr) [reply ramls/mod-users-bl.raml]: Makes sense. Thank you!

File: src/test/java/org/folio/rest/OpenTransactionsTest.java
Comment (rchr) [reply src/test/java/org/folio/rest/OpenTransactionsTest.java]: Thanks for pointing me to that!

File: src/test/java/org/folio/rest/OpenTransactionsTest.java
Comment (rchr) [reply src/test/java/org/folio/rest/OpenTransactionsTest.java:69]: You're right. Thank you!

File: src/main/java/org/folio/service/transactions/OpenTransactionsServiceImpl.java
Comment (julianladisch) [inline src/main/java/org/folio/service/transactions/OpenTransactionsServiceImpl.java:81]:
```
  @Override
  public Future<OpenTransactions> getTransactionsOfUser(User user, OkapiConnectionParams connectionParams) {
    String id = user.getId();
    final var openTransactions = new OpenTransactions().withUserId(id).withUserBarcode(user.getBarcode());

    Future<Integer> loansFuture =
      circulationClient.getOpenLoansCountByUserId(id, connectionParams)
        .onSuccess(numberLoans -> openTransactions.setLoans(numberLoans));

    Future<Integer> requestsFuture =
      circulationClient.getOpenRequestsCountByUserId(id, connectionParams)
        .onSuccess(numberReqs -> openTransactions.setRequests(numberReqs));

    Future<Integer> accountsFuture =
      feesFinesClient.getOpenAccountsCountByUserId(id, connectionParams)
        .onSuccess(accounts -> openTransactions.setFeesFines(accounts));

    Future<Integer> manualBlockFuture =
      feesFinesClient.getManualBlocksCountByUserId(id, connectionParams)
        .onSuccess(manualblocks -> openTransactions.setBlocks(manualblocks));

    Future<Integer> proxiesFuture =
      userClient.getProxiesCountByUserId(id, connectionParams)
        .onSuccess(p -> openTransactions.setProxies(p));

    return CompositeFuture.all(loansFuture, requestsFuture, accountsFuture, manualBlockFuture, proxiesFuture)
        .map(compositeFuture -> {
          boolean hasOpenTransactions = compositeFuture.result().list().stream()
              .map(o -> (Integer) o)
              .anyMatch(n -> n > 0);
          openTransactions.setHasOpenTransactions(hasOpenTransactions);
          return openTransactions;
        });
  }
```
Map.Entry stuff is not needed if each Future directly writes the result into openTransactions.
The Promise is not needed when directly returning CompositieFuture.all(...).map(...).
Calling setUserBarcode(null) or withUserBarcode(null) is fine and avoids the if clause.

File: ramls/openTransactions.json
Comment (julianladisch) [inline ramls/openTransactions.json:17]:
```suggestion
      "description": "Specifies if the user has any open transactions (loans, requests, fees/fines, proxies, blocks)"
```
Explain what "transactions" means.

File: src/main/java/org/folio/rest/client/impl/FeesFinesModuleClientImpl.java
Comment (julianladisch) [inline src/main/java/org/folio/rest/client/impl/FeesFinesModuleClientImpl.java:37]: `return response.getJson().getInteger("totalRecords");`
This avoids the JSON to Pojo mapping and allows to remove the accounts.json file and to remove the `accounts: !include accounts.json` line from mod-users-bl.raml.
The same for loans, requests, manualBlocks and proxyForCollection.

File: src/main/java/org/folio/service/transactions/OpenTransactionsServiceImpl.java
Comment (rchr) [reply src/main/java/org/folio/service/transactions/OpenTransactionsServiceImpl.java:57]: This is much cleaner. Thanks!

File: src/main/java/org/folio/rest/client/impl/FeesFinesModuleClientImpl.java
Comment (rchr) [reply src/main/java/org/folio/rest/client/impl/FeesFinesModuleClientImpl.java]: Makes sense and it's cleaner!

**Comment by rchr** [comment]
Test coverage now dropped to 79.4%. Do you think it is acceptable or should I write additional tests?

File: src/main/java/org/folio/rest/client/impl/CirculationStorageModuleClientImpl.java
Comment (julianladisch) [inline src/main/java/org/folio/rest/client/impl/CirculationStorageModuleClientImpl.java:49]:
```suggestion
    String requestUrl = connectionParams.getOkapiUrl() + "/request-storage/requests?limit=0&query=" + query;
```

File: src/main/java/org/folio/rest/client/impl/CirculationStorageModuleClientImpl.java
Comment (julianladisch) [inline src/main/java/org/folio/rest/client/impl/CirculationStorageModuleClientImpl.java:48]: duplicate urlEncode

File: src/test/java/org/folio/rest/JsonStoreTest.java
Comment (julianladisch) [inline src/test/java/org/folio/rest/JsonStoreTest.java:76]: SonarCloud complains
https://sonarcloud.io/project/issues?id=org.folio%3Amod-users-bl&pullRequest=116&resolved=false&types=CODE_SMELL
that assertTrue(... == ...) and assertTrue(...equals(...))
should be replaced.
You may use hamcreast that rest-assured ships with:
assertThat(..., is(...))

**Comment by julianladisch** [comment]
You may add a test when there are no optional dependencies:
```
  @Test
  public void getTransactionsNoOptionalDependencies(TestContext context) {
    int portOfVerticle = NetworkUtils.nextFreePort();
    DeploymentOptions usersBLOptions = new DeploymentOptions()
        .setConfig(new JsonObject().put("http.port", portOfVerticle));
    Async async1 = context.async();
    vertx.deployVerticle(RestVerticle.class, usersBLOptions)
    .onComplete(context.asyncAssertSuccess(x -> async1.complete()));
    async1.await(5000);

    int portWith404 = NetworkUtils.nextFreePort();
    Async async2 = context.async();
    vertx.createHttpServer()
    .requestHandler(request -> {
      System.out.println(request.path());
      if (request.path().startsWith("/users/")) {
        request.response().setStatusCode(200).end("{}");
        return;
      }
      request.response().setStatusCode(404).end();
    }).listen(portWith404)
    .onComplete(context.asyncAssertSuccess(x -> async2.complete()));
    async2.await(5000);

    given()
    .headers("X-Okapi-URL", "http://localhost:" + portWith404,
        "X-Okapi-Tenant", "supertenant",
        "X-Okapi-Token", token("supertenant", "maxi"))
    .port(portOfVerticle)
    .when()
    .get("/bl-users/by-id/" + USER_ID + "/open-transactions")
    .then()
    .statusCode(200)
    .body("hasOpenTransactions", equalTo(false),
      "loans", equalTo(0),
      "requests", equalTo(0),
      "feesFines", equalTo(0),
      "proxies", equalTo(0),
      "blocks", equalTo(0));
  }
```

File: src/main/java/org/folio/rest/client/impl/CirculationStorageModuleClientImpl.java
Comment (rchr) [reply src/main/java/org/folio/rest/client/impl/CirculationStorageModuleClientImpl.java]: Thank you for catching this!

File: src/main/java/org/folio/rest/client/impl/UserModuleClientImpl.java
Comment (julianladisch) [inline src/main/java/org/folio/rest/client/impl/UserModuleClientImpl.java:47]: The userName should be cql encoded, the query should be percent or urlencoded:
https://github.com/folio-org/raml-module-builder/blob/v32.2.2/util/src/main/java/org/folio/util/StringUtil.java#L65

```
String requestUrl = connectionParams.getOkapiUrl() + "/users?query="
  + StringUtil.urlEncode("username==" + StringUtil.cqlEncode(userName));
```

**Comment by rchr** [comment]
The build fails with error:
```
Tests in error: 
  BLUsersAPITest.before:81->insertData:100 » Connect Connection refused (Connect...
```
You fixed it in [PR-117](https://github.com/folio-org/mod-users-bl/pull/117), but it isn't merged yet. Should I wait until PR-117 is merged?

**Pull request merged**
